datastream-server/src/test/java/com/linkedin/datastream/server/zk/TestZkAdapter.java

```diff
@@ -715,9 +725,76 @@ ZkClient createZkClient() {
       return _zkClient;
     }
 
+    public class ZkClientMockStateChangeListener extends ZkStateChangeListener {
+      boolean sessionExpired = false;
+      @Override
+      public void handleStateChanged(Watcher.Event.KeeperState state) {
+        super.handleStateChanged(state);
+        if (state == Watcher.Event.KeeperState.Expired) {
+          LOG.info("ZkStateChangeListener::Session expired.");
+          sessionExpired = true;
+        }
+      }
+    }
+
+    @Override
+    ZkStateChangeListener getOrCreateStateChangeListener() {
+      _zkClientMockStateChangeListener = new ZkClientMockStateChangeListener();
+      return _zkClientMockStateChangeListener;
+    }
+
     public ZkClient getZkClient() {
       return _zkClient;
     }
+
+    public ZkClientMockStateChangeListener getZkStateChangeListener() {
+      return _zkClientMockStateChangeListener;
+    }
+  }
+
+  @Test
```

Comment:
Is is possible to simulate the "Disconnected" state? If so, can we add a test for that, and validate that scheduleExpiryTimerAfterSessionTimeout() is indeed called too? Then validate two cases, a) Let it expire and validate that exitOnSessionExpiryOrEstablishmentError() is called, b) Don't let it expire, and validate that exitOnSessionExpiryOrEstablishmentError()  is not called.

Reply:
No, I tried a lot and could not find a way. I couldn't find it in samza repo as well.